### PR TITLE
[WIP] [POC] Implement mechanism for functions without access to 'module' to still provide warnings

### PR DESCRIPTION
**Pull request by sivel**
##### SUMMARY
Implement mechanism for functions without access to 'module' to still provide warnings
##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
lib/ansible/module_utils/basic.py

##### ADDITIONAL INFORMATION

```paste below

```

### CODE REVIEW

**Comment by sivel** [comment]
bot_skip

**Review by nitzmahone — COMMENTED** [review]
Looks good- only case I can think of that could be weird is IIRC there are a handful of modules around that create AnsibleModule more than once, which could lead to some interesting re-entrant/order-dependent-freeing issues with the temporary warning filter... Not sure if there's a sane way to handle that globally, or if it's worth the trouble.

File: lib/ansible/module_utils/basic.py [hunk 1]
```diff
@@ -862,6 +867,12 @@ def __init__(self, argument_spec, bypass_checks=False, no_log=False,
                            ' Update the code for this module  In the future, AnsibleModule will'
                            ' always check for invalid arguments.', version='2.9')
 
+    def __del__(self):
+        try:
+            self._catch_warnings.__exit__(tuple())
```
Comment (webknjaz) [inline lib/ansible/module_utils/basic.py:872]: FTR in non-CPython implementations the time of when GC will call this may be inconsistent.

Comment (webknjaz) [reply lib/ansible/module_utils/basic.py:872]: Also, a tuple is not expected arg type: [https://docs.python.org/3/reference/datamodel.html#object.__exit__](https://docs.python.org/3/reference/datamodel.html#object.__exit__)

> If the context was exited without an exception, all three arguments will be None.


```suggestion
            self._catch_warnings.__exit__(None, None, None)
```

**Pull request closed without merging**